src/common/analytics.ts

```diff
@@ -1,17 +1,74 @@
 import * as applicationinsights from 'applicationinsights';
+import * as got from 'got';
 
+const apiEndpointHostname = process.env.API_ENDPOINT_HOSTNAME!; // eslint-disable-line no-process-env
 const instrumentationKey = process.env.RESULT_INSTRUMENTATION_KEY!; // eslint-disable-line no-process-env
 
 applicationinsights.setup(instrumentationKey).start();
 
 const appInsightsClient = applicationinsights.defaultClient;
+const options = {
+    batchDelay: 15000,
+    defaultProperties: {}
+};
+let sendTimeout: any = null;
+let telemetryQueue: any = [];
+
+const post = async () => {
+    try {
+        const response = await got.post(apiEndpointHostname, { json: telemetryQueue });
+
+        console.log(response.body);
+    } catch (error) {
+        console.warn(error.response.body);
+    }
+};
+
+const sendTelemetry = async () => {
+    if (sendTimeout) {
+        clearTimeout(sendTimeout);
+        sendTimeout = null;
+    }
+
+    try {
+        await post();
+        telemetryQueue = [];
+    } catch (err) {
+        console.warn('Failed to send telemetry: ', err);
+    }
```

Comment:
I think there's a possibility that we might not send the right data if the `post` takes too long, and might end up duplicating  or deleting data.

1. We call `sendTelemetry`
1. We get stuck on `post` because the server takes too long (I believe regular network timeouts are 30 seconds, so 15 more than the `batchDelay` we have)
1. `pushToQueue` gets called again and another `setTimeout` is set while we are still processing the previous one
1. Because we haven't yet cleared `telemetryQueue` we send it again in the second `sendTelemetry` call which is no good

Another scenario could be that the first `setTimeout` ends and clears all the contents of `telemetryQueue` so when the second `sendTelemetry` is finally invoked it doesn't contain any data and we've lost data.

Hope I've explained this correctly 😅

Reply:
Makes sense. Do you suggest that we drop the queue approach and post directly? And maybe let the client decide whether they want to batch their telemetry post, instead of this package? Perhaps, that would also make it more flexible, if it's a simple once a day type thing, we could simply call the trackEvent with the data and be done with it. It is much more frequent, the client can batch it up then call trackEvent with the queue itself. 

For the scenario of post failing, currently we don't retry. Do you think it's a good idea to add  backoff retry of some sort?

Reply:
I think the queue should remain here, don't know how often we are sending the telemetry in the different clients but maybe we can just pass the data to send to `post`? I think that something like the following should work.


```ts
const post = async (data: any[]) => {
    try {
        const response = await got.post(apiEndpointHostname, { json: data});

        console.log(response.body);
    } catch (error) {
        console.warn(error.response.body);
    }
};

const sendTelemetry = async () => {
    if (sendTimeout) {
        clearTimeout(sendTimeout);
        sendTimeout = null;
    }

    try {
        const data = telemetryQueue.splice(0);
        await post(data);
    } catch (err) {
        console.warn('Failed to send telemetry: ', err);
    }
```

RE the retry if `post` fails, maybe we can open an issue and tackle it at a later time if it becomes a problem?

Reply:
Ok this sounds good to me. I'll make the changes.